californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -396,11 +398,12 @@ public void completed(final Exchange exchange) {
 
 		@Override
 		public void contextEstablished(final Exchange exchange) {
-
-			if (exchange.getRequest() != null) {
-				observationStore.setContext(exchange.getRequest().getToken(), exchange.getCorrelationContext());
+			Request currentRequest = exchange.getCurrentRequest(); 
+			Request request = exchange.getRequest(); 
+			if (request == currentRequest && request != null && request.isObserve()) {
+				observationStore.setContext(request.getToken(), exchange.getCorrelationContext());
```

Comment:
Do you propose, to assume, that, if the correlation context is set once for an exchange, it cannot change? I don't agree, simply because I don't know all the cases where a exchange is used.

There is currently a ongoing discussion about the request response matching in the ietf core mailing list. Before there is a conclusion there, I would not like to fix the behaviour in that way.
(https://www.ietf.org/mail-archive/web/core/current/msg08814.html).

So let me point again to first discuss issue #311.
For me it's more important to know, which cases would require a setContext XyzStorage call. For now, it seems to be observe. Are there other use cases? Why do we need setContext in both storages? Why is one called for the request, the other one for currentRequest? For me, this seems to be wired enough, and I would first go for understanding the usage of "setContext" in the storages.